1-Draft/RFCNNNN-Supports-Parallel.md

```diff
@@ -0,0 +1,316 @@
+        # jobs to complete before continuing with end (and dispose) blocks.
+
+        # NEW! Unlike Start-Job commands where you must use the "using:" prefix
+        # on variables to access variables that are defined outside of the
+        # scope of the script block, all variables that are defined in the
+        # current scope would be accessible when the command is invoked with
+        # parallel processing without having to use the "using:" prefix. This
+        # is a smarter design for working with variables from other runspaces,
+        # and allows for easier transition of a command that does not support
+        # parallel processing into a command that does support parallel
+        # processing. The "using:" qualifier should only be used when you want
+        # to reference a variable that is outside of the scope of the local
+        # function.
+
+        # NEW! "using:" does not allow for transport of script blocks. This was
+        # done as a security measure (see [PowerShell Issue 9703](https://github.com/PowerShell/PowerShell/issues/9703) for details);
+        # however, there are valid cases where script blocks must be
+        # transportable into jobs, and this feature is one such case. Any
+        # variables that contain script blocks (i.e. parameters, variables
+        # defined in a begin block, etc.) must be made available to the thread
+        # jobs where process is invoked. Additionally, local functions defined
+        # in a begin block must be made available to the thread jobs where
+        # process is invoked. In general, local variables (from parameters or
+        # the begin block) and any resources that are created in the begin
+        # block such as functions, etc. must be made available in the runspaces
+        # where the threadjobs will run.
+    }
+    end {
+        # No changes here, this would be invoked synchronously in the current
+        # session.
+    }
+}
+
+$twoSessions = @($session1, $session2)
+Invoke-Command -Parallel -Session $twoSessions -ScriptBlock {
+    'This will be run in a thread job for each session.'
+}
+
+Invoke-Command -Session $twoSessions -ScriptBlock {
+    'This will be run once for each session, sequentially.'
+}
+```
+
+```output
+This will be run in a thread job for each session.
+This will be run in a thread job for each session.
+This will be run once for each session, sequentially.
+This will be run once for each session, sequentially.
+```
+
+## Specification
+
+For this to work, the following rough list of changes would have to be implemented:
+
+* update `CmdletBinding` to store the names of parameter sets that support
+parallel execution in the `PSCmdlet` object.
+* update the common parameter assignment to add `-Parallel`, `-ThreadLimit`,
+and `-TimeoutSecs` parameters to parameter sets that support `SupportsParallel`.
+* add logic to the command processor that identifies the local variables and
```

Comment:
>add logic to the command processor that identifies the local variables...

Can you explain what you mean by "local variables".  PowerShell doesn't really have the concept of "local variables".  Rather, it employs (among other things) a multitude of nested scopes hosted in SessionStates that may or may not themselves be related.  AFAICT each variable encountered during scriptblock invocation by following a search algorithm.  In general, I don't think something like "identifying the local variables" is in general computable.  For example, in

```PowerShell

$g = Get-SomeScriptBlockGenerator

1..10 | Invoke-Command { . $g.GetNextScriptBlock() }
```

The SessionState (and therefore variables in scope) of the scriptblock returned by .GetNextScriptBlock() cannot be discovered prior to call the .GetNextScriptBlock().


There is another way, I think.  AFAICT the search "rules" for a runspace could be expanded to include the SessionState of the call site of Invoke-Command of the parent runspace.